loopy/target/pyopencl.py

```diff
@@ -244,18 +244,30 @@ def map_sum(self, expr, type_context):
 
             c_applied = [self.rec(c, type_context, tgt_dtype) for c in complexes]
 
+            mul_name = f"{tgt_name}_mul"
+
             def binary_tree_add(start, end):
                 if start + 1 == end:
                     return c_applied[start]
                 mid = (start + end)//2
                 lsum = binary_tree_add(start, mid)
                 rsum = binary_tree_add(mid, end)
-                return p.Variable("%s_add" % tgt_name)(lsum, rsum)
+
+                if isinstance(lsum, p.Call) and isinstance(lsum.function,
+                        p.Variable) and lsum.function.name == mul_name:
+                    return p.Variable(f"{tgt_name}_fma")(*lsum.parameters, rsum)
+
+                elif isinstance(rsum, p.Call) and isinstance(rsum.function,
+                        p.Variable) and rsum.function.name == mul_name:
+                    return p.Variable(f"{tgt_name}_fma")(*rsum.parameters, lsum)
+
+                else:
+                    return p.Variable(f"{tgt_name}_add")(lsum, rsum)
```

Comment:
It's mildly odd that we're doing this for complex but not for reals.

Reply:
For reals, we generate code that looks like `res = c + a * b`.
For complex, we previously generated code like `complex_add(c, complex_mul(a, b))`.
This generated code like below.
```
tmp.real = a.real * b.real - a.imag * b.imag
tmp.imag = a.real * b.imag + a.imag * b.real
res.real = c.real + tmp.real
res.imag = c.imag + tmp.imag
```
The compiler generated 2 fma whereas with `cdouble_fma`, the compiler generated 4 fmas.

I've seen this in other places where introducing temporaries prevented generating fmas (eg: https://github.com/inducer/sumpy/pull/112)
clang seems to not be able to fuse across statements even though `-ffp-contract=fast` is the default for the CUDA target.

Reply:
Could you add comments to the code explaining these aspects?

Reply:
Done in 87a19b15